calendra/europe/united_kingdom.py

```diff
@@ -18,26 +17,72 @@ class UnitedKingdom(WesternCalendar, ChristianMixin):
+        1973: [(date(1973, 11, 14), "Royal wedding"), ],
+        1977: [(date(1977, 6, 7), "Queen’s Silver Jubilee"), ],
+        1981: [(date(1981, 7, 29), "Royal wedding"), ],
+        1999: [(date(1999, 12, 31), "New Year's Eve"), ],
+        2002: [(date(2002, 6, 3), "Queen’s Golden Jubilee"), ],
+        2011: [(date(2011, 4, 29), "Royal Wedding"), ],
+        2012: [(date(2012, 6, 5), "Queen’s Diamond Jubilee"), ],
+    }
 
-    def get_variable_days(self, year):
-        days = super(UnitedKingdom, self).get_variable_days(year)
-        days += [
-            Holiday(
-                date(year, 5, 1) + rd.relativedelta(weekday=rd.MO(1)),
+    def get_early_may_bank_holiday(self, year):
+        """
+        Return Early May bank holiday
+        """
+        # Special case in 2020, for the 75th anniversary of the end of WWII.
+        if year == 2020:
+            return Holiday(
+                date(year, 5, 8),
+                "Early May bank holiday (VE day)",
+                indication="VE day",
+            )
+        else:
+            return Holiday(
+                UnitedKingdom.get_nth_weekday_in_month(year, 5, MON),
                 "Early May Bank Holiday",
                 indication="1st Monday in May",
-            ),
-            Holiday(
-                date(year, 5, 30) + rd.relativedelta(weekday=rd.MO(-1)),
-                "Spring Bank Holiday",
-                indication="Last Monday in May",
-            ),
-            Holiday(
-                date(year, 8, 31) + rd.relativedelta(weekday=rd.MO(-1)),
-                "Late Summer Bank Holiday",
-                indication="Last Monday in August",
-            ),
-        ]
+            )
+
+    def get_spring_bank_holiday(self, year):
+        if year == 2012:
+            spring_bank_holiday = date(2012, 6, 4)
+        elif year == 1977:
+            spring_bank_holiday = date(1977, 6, 6)
+        elif year == 2002:
+            spring_bank_holiday = date(2002, 6, 4)
+        else:
+            spring_bank_holiday = UnitedKingdom. \
+                get_last_weekday_in_month(year, 5, MON)
+        return Holiday(
+            spring_bank_holiday,
+            "Spring Bank Holiday",
+            indication="Last Monday in May"
+        )
+
+    def get_late_summer_bank_holiday(self, year):
+        return Holiday(
+            UnitedKingdom.get_last_weekday_in_month(year, 8, MON),
+            "Late Summer Bank Holiday",
+            indication="Last Monday in August",
+        )
+
+    def non_computable_holiday(self, year):
+        non_computable = self.non_computable_holiday_dict.get(year, None)
+        return non_computable
+
+    def get_variable_days(self, year):
+        days = super(UnitedKingdom, self).get_variable_days(year)
+        days.append(self.get_early_may_bank_holiday(year))
+        days.append(self.get_spring_bank_holiday(year))
+        days.append(self.get_late_summer_bank_holiday(year))
+        # Boxing day & XMas shift
```

Comment:
Boxing day and XMas shifts are handled by the Holiday class. Here's where the real value of Calendra comes in - instead of defining these holidays in duplicate, the holidays are able to shift on their own.

Reply:
Probably you or I should add a test that ensures that the shift days are excluded from the list of holidays.

Reply:
Again, this is how upstream does it. But if you can perhaps illustrate how you'd prefer it done, I can do the legwork here and elsewhere where the same pattern applies.

Reply:
Right - upstream is sloppy. This is the key difference between calendra and workalendar. In workalendar, when a holiday is indicated on one day and observed on another, the author just creates two "holidays". Calendra is smarter and models the holiday as a single Holiday object with 'indicated' and 'observed' dates (depending on the context). It's this technique that allows one to correctly count the number of holidays and to present to a consumer the indicated or observed date based on context.

I'll tackle UK and things should become clearer.

Reply:
It occurs to me a good way to see what changes are intended by calendra is to compare the [last merge with workalendar in that merge](https://github.com/jaraco/calendra/compare/1101b41347e6e80a07f35382cd2fd14b79912d4a...39e66d2894332cc544ade87085fbcdb5ae7be5e4#diff-bd335d1ced2eba5fdd5042459b93c201). That link shows the intended differences between calendra and workalendar when calendra 3.2.1 was released. I'll mention that in #8 also.